plugins/BEdita/API/src/Auth/OTPAuthenticate.php

```diff
@@ -0,0 +1,205 @@
+     * - `fields` The fields to use to identify a user by.
+     * - `userModel` The alias for users table, defaults to Users.
+     * - `finder` The finder method to use to fetch user record. Defaults to 'all'.
+     *   You can set finder name as string or an array where key is finder name and value
+     *   is an array passed to `Table::find()` options.
+     *   E.g. ['finderName' => ['some_finder_option' => 'some_value']]
+     *  - `passwordHasher` Password hasher class. Can be a string specifying class name
+     *    or an array containing `className` key, any other keys will be passed as
+     *    config to the class. Defaults to 'Default'.
+     * - 'expiry' Expiry time of a user token, expressed as string expression like `+1 hour`, `+10 minutes`
+     * - 'generator' Secret token generator, if a valid callable is used instead of default one.
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'authProviders' => [],
+        'fields' => [
+            'username' => 'username',
+            'password' => null,
+        ],
+        'userModel' => 'Users',
+        'finder' => 'all',
+        'contain' => null,
+        'passwordHasher' => 'Default',
+        'expiry' => '+15 minutes',
+        'generator' => null,
+    ];
+
+    /**
+     * {@inheritDoc}
+     */
+    public function __construct(ComponentRegistry $registry, array $config = [])
+    {
+        parent::__construct($registry, $config);
+        // override configuration with `otp` auth provider params ('generator', 'expiry'...)
+        $authProvider = $this->getConfig('authProviders.otp');
+        if ($authProvider && $authProvider instanceof AuthProvider) {
+            $this->setConfig((array)$authProvider->get('params'), true);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function authenticate(ServerRequest $request, Response $response)
+    {
+        $username = $request->getData('username');
+        if (empty($username) || !is_string($username)) {
+            return false;
+        }
+
+        $grant = $request->getData('grant_type');
+        if ($grant === 'otp') {
+            return $this->otpAccess($username, $request);
+        } elseif ($grant === 'otp_request') {
+            return $this->otpRequest($username);
+        }
+
+        return false;
+    }
+
+    /**
+     * Retrieve access grant using authorization code and secret token.
+     *
+     * @param string $username User name
+     * @param ServerRequest $request Request object
+     * @return array|bool User data array on success, false on failure
+     */
+    protected function otpAccess($username, ServerRequest $request)
+    {
+        $result = $this->_findUser($username);
+        if (empty($result)) {
+            return false;
+        }
+
+        if (empty($request->getData('authorization_code')) || empty($request->getData('token'))) {
```

Comment:
[minor] I would move this check before of  `_findUser()` to avoid useless query on db if missing required params.